internal/pkg/agent/application/actions/handlers/handler_action_unenroll.go

```diff
@@ -40,18 +54,61 @@ type Unenroll struct {
 // NewUnenroll creates a new Unenroll handler.
 func NewUnenroll(
 	log *logger.Logger,
+	coord ActionCoordinator,
 	ch chan coordinator.ConfigChange,
 	closers []context.CancelFunc,
 	stateStore stateStore,
 ) *Unenroll {
 	return &Unenroll{
 		log:        log,
+		coord:      coord,
 		ch:         ch,
 		closers:    closers,
 		stateStore: stateStore,
 	}
 }
 
+type performActionFunc func(context.Context, component.Component, component.Unit, string, map[string]interface{}) (map[string]interface{}, error)
+
+func dispatchUnenrollInParallel(ctx context.Context, log *logp.Logger, action *fleetapi.ActionUnenroll, comps []component.Component, units []component.Unit, performAction performActionFunc) error {
+	if action == nil {
+		return nil
+	}
+
+	// Deserialize the action into map[string]interface{} for dispatching over to the apps
+	params, err := action.MarshalMap()
+	if err != nil {
+		return err
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	// Iterate through found components and forward the UNENROLL action
+	for i := 0; i < len(comps); i++ {
+		g.Go(func(idx int) func() error {
+			comp := comps[i]
+			unit := units[i]
+			return func() error {
+				log.Debugf("Dispatch %v action to %v", action.Type(), unit.Config.Type)
+				res, err := performAction(ctx, comp, unit, unit.Config.Type, params)
+				if err != nil {
+					log.Warnf("UNENROLL failed to dispatch to %v, err: %v", comp.ID, err)
+					return err
+				} else {
+					strErr := readMapString(res, "error", "")
+					if strErr != "" {
+						log.Warnf("UNENROLL failed for %v, err: %v", comp.ID, strErr)
+						return errors.New(strErr)
+					}
+				}
+				return nil
+			}
+		}(i))
+	}
+
+	return g.Wait()
```

Comment:
In my opinion it's appropriate to run the actions dispatch in parallel like this, this would return the first error and cancel context on the first error returned by any of the "Go routined" functions. This pattern allows early cancellation on error.
The current agent implementation has fairly few places that use the same ```errgroup``` approach. 
Also if you insist on returning the array of errors then the more context would need to be attached to every error in order to be able to correlate these errors to specific dispatch call, because the order of the errors will be random in general.

Please let me know if you feel strongly about returning an array of errors or the wrapper error containing that array instead and still want me to change the code.